README.md

```diff
@@ -46,41 +70,48 @@ end
 
 and never use `def change`. We believe that this is the only safe approach in production environments. For development environments we iterate by recreating the database from scratch every time we make a change.
 
-### Migrations
-
-There are two major classes of concerns we try to handle in the API:
-
-- Database safety (e.g., long-held locks)
-- Application safety (e.g., dropping columns the app uses)
+### Transactional DDL
 
-We rename migration methods with prefixes denoting their safety level:
+Individual DDL statements in PostgreSQL are transactional by default (as are all Postgres statements). Concurrent index creation and removal are two exceptions: these utility commands manage their own transaction state (and each uses multiple transactions to achieve the desired concurrency).
 
-- `safe_*`: These methods check for both application and database safety concerns, prefer concurrent operations where available, set low lock timeouts where appropriate, and decompose operations into multiple safe steps.
-- `unsafe_*`: These methods are generally a direct dispatch to the native ActiveRecord migration method.
+We [disable ActiveRecord's DDL transactions](./lib/pg_ha_migrations/hacks/disable_ddl_transaction.rb) (which wrap the entire migration file in a transaction) by default for the following reasons:
 
-Calling the original migration methods without a prefix will raise an error.
+* [Running multiple DDL statements inside a transaction acquires exclusive locks on all of the modified objects](https://medium.com/paypal-tech/postgresql-at-scale-database-schema-changes-without-downtime-20d3749ed680#cc22).
+* Acquired locks are held until the end of the transaction.
+* Multiple locks creates the possibility of deadlocks.
+* Increased exposure to long waits:
+  * Each newly acquired lock has its own timeout applied (so total lock time is additive).
+  * [Safe lock acquisition](#safely_acquire_lock_for_table) (which is used in each migration method where locks will be acquired) can issue multiple lock attempts on lock timeouts (with sleep delays between attempts).
 
-The API is designed to be explicit yet remain flexible. There may be situations where invoking the `unsafe_*` method is preferred (or the only option available for definitionally unsafe operations).
+You can change this by resetting `ActiveRecord::Migration.disable_ddl_transaction` in your application.
```

Comment:
I don't think we have any tests for this, so I'm not sure about leaving it in here, but also it was already present, so I wasn't sure if we should yank it in this change.

Reply:
I did some testing with this and I don't think it even works. `disable_ddl_transaction` is a class instance variable and with the way inheritance works in Ruby, it isn't propagated to individual migration classes that inherit from `ActiveRecord::Migration`.

```ruby
# ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [x86_64-linux]
?> class Parent
?>   class << self
?>     attr_accessor :foo
?>   end
>> end
=> [:foo, :foo=]
>> Parent.foo = "bar"
=> "bar"
>> Parent.foo
=> "bar"
?> class Child < Parent
>> end
=> nil
>> Child.foo
=> nil
```

So until https://github.com/braintree/pg_ha_migrations/pull/96 is merged, I don't think we have a way to globally change the DDL transaction behavior. Of course, you should still be able reset this in each individual migration class with:

```ruby
class MyTransactionalMigration < ActiveRecord::Migration[7.1]
  self.disable_ddl_transaction = false

  def up
    # do some stuff
  end
end
```